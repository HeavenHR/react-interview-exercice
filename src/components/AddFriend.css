:local(.addFriendContainer) {
  margin: 10px;
}

:local(.addFriendButtonHolder) {
  margin-bottom: 10px;
}

:local(.tickbox) {
  margin: 10px; 
  display: inline;
}

:local(.color) {
  color: lightgreen;
}
