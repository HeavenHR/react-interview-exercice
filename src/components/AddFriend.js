import React, { Component, PropTypes } from "react";
import classnames from "classnames";

import AddFriendInput from "./AddFriendInput";
import AddFriendRadio from "./AddFriendRadio";

import { MALE } from "./../constants/PageTypes";

import styles from "./AddFriend.css";

class AddFriend extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      name: "",
      gender: MALE
    };
  }

  render() {
    return (
      <div onKeyDown={this.handleKeyDown} className={styles.addFriendContainer}>
        <AddFriendInput name={this.state.name} onHandleChange={this.handleChange} />
        <AddFriendRadio gender={this.state.gender} onHandleChange={this.handleChange} />
        <div className={classnames("text-center", styles.addFriendButtonHolder)}>
          <button
            type='button'
            className='btn'
            onClick={this.handleSubmit}
            disabled={this.state.name.length > 0 ? false : true}>
            Add Friend
          </button>
        </div>
      </div>
    );
  }

  resetState = () => {
    this.setState({ name: "", gender: MALE });
  };

  handleChange = (key, value) => {
    this.setState({ [key]: value });
  };

  handleSubmit = () => {
    const nameFromState = this.state.name.trim();
    this.props.addFriend({ name: nameFromState, gender: this.state.gender });
    this.resetState();
  };

  handleKeyDown = (e) => {
    const nameFromState = this.state.name.trim();
    if (e.which === 13 && nameFromState.length > 0) {
      this.handleSubmit();
    }
  };
}

AddFriend.propTypes = {
  addFriend: PropTypes.func.isRequired
};

export default AddFriend;
