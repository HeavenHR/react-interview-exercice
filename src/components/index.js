export { default as AddFriend } from './AddFriend';
export { default as AddFriendInput } from './AddFriendInput';
export { default as ButtonGroup } from './ButtonGroup';
export { default as FriendList } from './FriendList';
export { default as FriendListItem } from './FriendListItem';
export { default as Pagination } from './Pagination';
export { default as ErrorMessage } from './ErrorMessage';
