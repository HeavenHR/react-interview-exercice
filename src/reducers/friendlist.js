import {random} from 'lodash'
import * as types from '../constants/ActionTypes';

const initialState = {
  friendsById: [
    {
      id:random(10000,1000000,false),
      name: 'Theodore Roosevelt',
      gender: 'male',
      starred: true
    },
    {
      id:random(10000,1000000,false),
      name: 'Abraham Lincoln',
      gender: 'male',
      starred: false
    },
    {
      id:random(10000,1000000,false),
      name: 'George Washington',
      gender: 'male',
      starred: false
    }
  ]
};

export default function friends(state = initialState, action) {
  switch (action.type) {
    case types.ADD_FRIEND:
      return {
        ...state,
        friendsById: [
          ...state.friendsById,
          {
            name: action.friend.name,
            gender: action.friend.gender,
            id:random(10000,1000000,false),
          }
        ],
      };
    case types.DELETE_FRIEND:
      return {
        ...state,
        friendsById: state.friendsById.filter((item) => item.id !== action.id)
      };
    case types.STAR_FRIEND:
      let friends = [...state.friendsById];
      let friend = friends.find((item) => item.id === action.id);
      friend.starred = !friend.starred;
      return {
        ...state,
        friendsById: friends
      };

    default:
      return state;
  }
}
