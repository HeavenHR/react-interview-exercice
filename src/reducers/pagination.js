import * as types from '../constants/ActionTypes';

const initialState = {
  currentPageNo: 0,
  pageSize: 2
};

export default function pagination(state = initialState, action) {
  switch (action.type) {

    case types.UPDATE_AFTER_CHANGE:
      const {currentPageNo} = action;
      return {
        ...state,
        currentPageNo
      };

    case types.SET_PAGE:
      return {
        ...state,
        currentPageNo: action.pageNo
      };

    default:
      return state;
  }
}
