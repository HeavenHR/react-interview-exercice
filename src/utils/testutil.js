import { mount, shallow } from 'enzyme';
import { prettyPrint } from 'html';
import * as types from '../constants/ActionTypes';

export const mockFunction = jest.fn(value => value);

export const sampleFriend = {
  name: 'Sample friend',
  gender: 'male'
};

export const sampleId = 1;

export const createMountedComponent = ({ children, isShallow }) => {
  const mountedComponent = isShallow ? shallow(children) : mount(children);
  mountedComponent.htmlRender = () => prettyPrint(mountedComponent.html(), { indent_size: 2 });

  return mountedComponent;
};

export const inputProps = {
  value: 'sample name',
  onSubmit: mockFunction,
  onChange: mockFunction,
  isInvalid: false,
  errorMessage: 'Please fill friend name'
};

export const buttonGroupProps = {
  selected: '',
  options: [
    { value: 'male', showIcon: true },
    { value: 'female', showIcon: true }
  ],
  onClick: mockFunction,
  isInvalid: false,
  errorMessage: 'Please select gender'
};

export const changeEventObject = { target: { value: 'sample text' } };

export const samplePages = [1, 2, 3, 4, 5, 6, 7];

export const totalPages = samplePages.length;

export const addFriendAction = {
  type: types.ADD_FRIEND,
  friend: sampleFriend
};

export const expectedAddFriendState = {
  friendsById: [
    {
      name: 'Sample friend',
      gender: 'male'
    },
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const deleteFriendAction = {
  type: types.DELETE_FRIEND,
  id: 0
};

export const expectedDeleteFriendState = {
  friendsById: [
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const starFriendAction = {
  type: types.STAR_FRIEND,
  id: 1
};

export const expectedStarFriendState = {
  friendsById: [
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: true,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const updatePaginationAction = {
  type: types.UPDATE_PAGINATION,
  currentPage: 2
};

export const expectedUpdatedPaginationState = {
  pageSize: 2,
  currentPage: 2,
  pagingRange: 3
};
