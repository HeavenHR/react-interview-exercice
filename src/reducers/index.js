import { combineReducers } from 'redux';
import { default as friendlist } from './friendlist';
import { pagination } from './pagination';

export default combineReducers({ friendlist, pagination });
