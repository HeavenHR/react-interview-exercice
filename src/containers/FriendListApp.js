import React, { Component } from 'react';
import styles from './FriendListApp.css';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import * as actionsObject from '../actions';
import { FriendList, Pagination, AddFriend } from '../components';
import { derivePageItems } from '../utils/pagination';

class FriendListApp extends Component {
  render() {
    const { friendlist: { friendsById }, pagination: { pageSize, currentPage, pagingRange }, actions } = this.props;
    const currentList = derivePageItems(friendsById, currentPage, pageSize) || []; // friendList should render based on currentPage
    const totalPages = Math.ceil(friendsById.length / pageSize);

    return (
      <div className={styles.friendListApp}>
        <h1>The FriendList</h1>
        <div className={styles.wrapper}>
          <AddFriend addFriend={actions.addFriend} />
          <FriendList friends={currentList} actions={actions} />
        </div>
        {totalPages > 1 &&
          <Pagination
            totalPages={totalPages}
            currentPage={currentPage}
            pagingRange={pagingRange}
            {...actions}
          />}
      </div>
    );
  }
}

function mapStateToProps(state) {
  return state
};

const mapDispatchToProps = dispatch => ({ actions: bindActionCreators(actionsObject, dispatch) });

export default connect(mapStateToProps, mapDispatchToProps)(FriendListApp)
