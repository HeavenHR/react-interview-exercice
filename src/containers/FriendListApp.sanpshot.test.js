import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';

import FriendListApp from './FriendListApp';
import { default as rootReducer } from '../reducers';
import { initialState as friendlist } from '../reducers/friendlist';
import { initialState as pagination } from '../reducers/pagination';
import { createMountedComponent } from '../utils/testutil';
import { FriendList, Pagination, AddFriend } from '../components';

const store = createStore(rootReducer);

describe('<FriendListApp /> should', () => {
  it('render correctly with store', () => {
    const mountedFriendListApp = createMountedComponent({
      children: (
        <Provider store={store}>
          <FriendListApp />
        </Provider>
      )
    });

    expect(mountedFriendListApp.htmlRender()).toMatchSnapshot();
    expect(mountedFriendListApp.props().store.getState()).toEqual({ friendlist, pagination });
    expect(mountedFriendListApp.find(FriendList)).toHaveLength(1);
    expect(mountedFriendListApp.find(AddFriend)).toHaveLength(1);
    expect(mountedFriendListApp.find(Pagination)).toHaveLength(1);
    expect(mountedFriendListApp.find('h1').contains('The FriendList')).toBeTruthy();
  });
});