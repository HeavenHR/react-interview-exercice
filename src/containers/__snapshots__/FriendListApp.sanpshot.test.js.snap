exports[`<FriendListApp /> should render correctly with store 1`] = `
"<div>
  <h1>The FriendList</h1>
  <div>
    <div class=\"addFriend\">
      <div class=\"form-group\">
        <input type=\"text\" class=\"form-control\" placeholder=\"Type the name of a friend\"
        value=\"\">
      </div>
      <div class=\"form-group\">
        <div class=\"btn-group btn-block\">
          <button type=\"button\" value=\"male\" class=\"btn btn-primary col-sm-6\"><i class=\"fa fa-male\"></i>
            <!-- react-text: 11 -->male
            <!-- /react-text -->
          </button>
          <button type=\"button\" value=\"female\" class=\"btn btn-primary col-sm-6\"><i class=\"fa fa-female\"></i>
            <!-- react-text: 14 -->female
            <!-- /react-text -->
          </button>
        </div>
      </div>
    </div>
    <ul class=\"list-group undefined\">
      <li class=\"list-group-item undefined\">
        <div>
          <div><span>Theodore Roosevelt</span>
          </div>
          <div><i class=\"fa fa-male\"></i><small>male</small>
          </div>
          <div><small>xx friends in common</small>
          </div>
        </div>
        <div>
          <button class=\"btn btn-default undefined\"><i class=\"fa fa-star\"></i>
          </button>
          <button class=\"btn btn-default undefined\"><i class=\"fa fa-trash\"></i>
          </button>
        </div>
      </li>
      <li class=\"list-group-item undefined\">
        <div>
          <div><span>Abraham Lincoln</span>
          </div>
          <div><i class=\"fa fa-male\"></i><small>male</small>
          </div>
          <div><small>xx friends in common</small>
          </div>
        </div>
        <div>
          <button class=\"btn btn-default undefined\"><i class=\"fa fa-star-o\"></i>
          </button>
          <button class=\"btn btn-default undefined\"><i class=\"fa fa-trash\"></i>
          </button>
        </div>
      </li>
    </ul>
  </div>
  <div class=\"\">
    <ul class=\"pagination\">
      <li class=\"page-item disabled\"><a class=\"page-link\" href=\"#\">&lt; prev</a>
      </li>
      <li class=\"page-item current\"><a class=\"page-link\" href=\"#\">1</a>
      </li>
      <li class=\"page-item\"><a class=\"page-link\" href=\"#\">2</a>
      </li>
      <li class=\"page-item\"><a class=\"page-link\" href=\"#\">next &gt;</a>
      </li>
    </ul>
  </div>
</div>"
`;
